File: ios/SecretaryVaultAccess/Sources/SecretaryVaultAccess/SyncCoordinator.swift
import Foundation

/// Orchestrates the two-call inspect→commit sync round-trip over a
/// `VaultSyncPort`. An `actor` so its state mutations are isolated. Note that
/// actors are reentrant across `await`: a second `runPass`/`resolve` can
/// interleave at the suspension point, so this type assumes a single serial
/// driver per vault. The freshness token committed by `resolve` is captured
/// before the suspension and is therefore never torn; a genuinely concurrent
/// commit is additionally backstopped on the FFI side by the per-vault lockfile
/// (surfacing `.inProgress`) and the `evidenceStale` freshness gate.
///
/// Secret hygiene: the password is passed per call and never stored. Only the
/// non-secret freshness token + veto metadata persist between calls.
public actor SyncCoordinator {
    private let port: VaultSyncPort
    private let stateDir: String
    private let vaultFolder: String

    /// Stashed across the round-trip after a paused `runPass`. The token is held
    /// here (private) and replayed by `resolve`; the public `pendingConflict`
    /// exposes only the display detail.
    private var stashedToken: [UInt8]?
    private var stashedConflict: PendingConflict?

    public init(port: VaultSyncPort, stateDir: String, vaultFolder: String) {
        self.port = port
        self.stateDir = stateDir
        self.vaultFolder = vaultFolder
    }

    /// The pending conflict detail from the last paused `runPass`, if any.
    public var pendingConflict: PendingConflict? { stashedConflict }

    /// Read-only sync status.
    public func status(vaultUuid: [UInt8]) async throws -> SyncStatus {
        try await port.status(stateDir: stateDir, vaultUuid: vaultUuid)
    }

    /// Run one inspect pass. On `.conflictsPending` the detail + token are
    /// stashed for a subsequent `resolve`; every other arm clears any prior stash.
    public func runPass(password: [UInt8], nowMs: UInt64) async throws -> SyncOutcome {
        let outcome = try await port.sync(stateDir: stateDir, vaultFolder: vaultFolder,
                                          password: password, nowMs: nowMs)
        stash(from: outcome)
        return outcome
    }

    /// Commit `decisions` against the stashed freshness token. Throws
    /// `.noPendingConflict` if `runPass` did not pause on a conflict.
    ///
    /// On a non-`conflictsPending` result the stash is cleared. On a thrown
    /// error (e.g. `.evidenceStale`) the stash is preserved so the caller can
    /// re-inspect and retry. If the recompute re-raises a conflict, the new
    /// detail replaces the old.
    public func resolve(decisions: [SyncVetoDecision],
                        password: [UInt8], nowMs: UInt64) async throws -> SyncOutcome {
        guard let token = stashedToken else { throw VaultSyncError.noPendingConflict }
        let outcome = try await port.commitDecisions(
            stateDir: stateDir, vaultFolder: vaultFolder, password: password,
            decisions: decisions, manifestHash: token, nowMs: nowMs)
        stash(from: outcome)
        return outcome
    }

    /// Update the stash from a pass outcome: keep the token + detail on a
    /// conflict, clear both on any resolved arm.
    private func stash(from outcome: SyncOutcome) {
        switch outcome {
        case let .conflictsPending(vetoes, collisions, manifestHash):
            stashedToken = manifestHash
            stashedConflict = PendingConflict(vetoes: vetoes, collisions: collisions)
        default:
            stashedToken = nil
            stashedConflict = nil
        }
    }
}

File: android/vault-access/src/main/kotlin/org/secretary/sync/SyncCoordinator.kt
package org.secretary.sync

import kotlinx.coroutines.sync.Mutex
import kotlinx.coroutines.sync.withLock

/**
 * Threads the two-call inspect→commit sync round-trip and holds the freshness token +
 * conflict detail privately between the two calls. One coordinator drives one vault.
 *
 * Concurrency: guarded by a non-reentrant [Mutex] held ACROSS the suspending port call.
 * This intentionally diverges from the iOS reentrant `actor`: a second concurrent
 * `runPass`/`resolve` blocks until the first completes rather than interleaving — stronger
 * (non-interleaving) serialization. It cannot deadlock because the public methods never
 * call one another. The per-vault FFI lockfile (surfaced as [VaultSyncError.InProgress])
 * remains the cross-process guard; this [Mutex] is the in-process single-driver guarantee.
 * All four methods share the mutex, so a [status]/[pendingConflict] read blocks behind an
 * in-flight pass — intended, since one coordinator drives one vault serially.
 *
 * Secret hygiene: the password is forwarded to the port per call and never retained. Only
 * the manifest-hash freshness token (not a secret) and conflict METADATA are stashed.
 */
class SyncCoordinator(
    private val port: VaultSyncPort,
    private val stateDir: String,
    private val vaultFolder: String,
) {
    private val mutex = Mutex()
    private var stashedToken: ByteArray? = null
    private var stashedConflict: PendingConflict? = null

    /** The conflict detail of a currently-paused pass, or null if none is stashed. */
    suspend fun pendingConflict(): PendingConflict? = mutex.withLock { stashedConflict }

    /** Read-only device-clock status. */
    suspend fun status(vaultUuid: ByteArray): SyncStatus =
        mutex.withLock { port.status(stateDir, vaultUuid) }

    /**
     * Run one inspect pass. On [SyncOutcome.ConflictsPending] the token + conflict are
     * stashed for [resolve]; every other arm clears any prior stash.
     */
    suspend fun runPass(password: ByteArray, nowMs: ULong): SyncOutcome = mutex.withLock {
        val outcome = port.sync(stateDir, vaultFolder, password, nowMs)
        applyStash(outcome)
        outcome
    }

    /**
     * Commit veto decisions for the paused pass, replaying the stashed freshness token.
     * Throws [VaultSyncError.NoPendingConflict] if nothing is stashed. A resolved arm clears
     * the stash; another [SyncOutcome.ConflictsPending] re-stashes the new token; a thrown
     * error (e.g. [VaultSyncError.EvidenceStale]) propagates and PRESERVES the stash so the
     * caller can retry `resolve` without a fresh `runPass`.
     */
    suspend fun resolve(
        decisions: List<SyncVetoDecision>,
        password: ByteArray,
        nowMs: ULong,
    ): SyncOutcome = mutex.withLock {
        val token = stashedToken ?: throw VaultSyncError.NoPendingConflict
        val outcome = port.commitDecisions(stateDir, vaultFolder, password, decisions, token, nowMs)
        applyStash(outcome)
        outcome
    }

    /** Stash on a paused pass; clear on any resolved/safe arm. Not called on a thrown error. */
    private fun applyStash(outcome: SyncOutcome) {
        if (outcome is SyncOutcome.ConflictsPending) {
            stashedToken = outcome.manifestHash
            stashedConflict = PendingConflict(outcome.vetoes, outcome.collisions)
        } else {
            stashedToken = null
            stashedConflict = null
        }
    }
}
